### PR TITLE
Use prometheus client instead of faraday

### PR DESCRIPTION
**Pull request by yaacov**
**Description**

We want to leverage the new Prometheus API gem, for metrics collection.

BZ:
https://bugzilla.redhat.com/show_bug.cgi?id=1468251 - Use Prometheus Metrics Ruby client

Note: this is the second part of https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/132

### CODE REVIEW

**Comment by yaacov** [comment]
@miq-bot add_label bug, metrics, gaprindashvili/yes

@cben @moolitayer @Ladas @agrare please review

BZ traget 5.9

File: app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/prometheus_client_mixin.rb [hunk 1]
```diff
@@ -10,22 +10,10 @@ def prometheus_client
...
-    Faraday.new(
-      :url     => uri.to_s,
-      :proxy   => options[:http_proxy_uri].empty? ? nil : options[:http_proxy_uri],
-      :ssl     => {
-        :verify     => options[:verify_ssl] != OpenSSL::SSL::VERIFY_NONE,
-        :cert_store => options[:ssl_cert_store]
-      },
-      :request => {
-        :open_timeout => worker_class.worker_settings[:prometheus_open_timeout] || 5,
-        :timeout      => worker_class.worker_settings[:prometheus_request_timeout] || 30
```
Comment (enoodle) [inline app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/prometheus_client_mixin.rb]: It seems like those timeouts are not  sent to the prometheus client anymore, are we OK with that?

File: app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/prometheus_client_mixin.rb [hunk 2]
```diff
@@ -59,17 +50,11 @@ def labels_to_s(labels, job = "kubernetes-cadvisor")
...
-      response = prometheus_client.get("query", :query => "ALL")
+      data = prometheus_client.query(:query => "ALL")
     rescue StandardError => err
       raise MiqException::MiqUnreachableError, err.message, err.backtrace
     end
 
-    begin
-      data = JSON.parse(response.body)
-    rescue StandardError => err # if auth_proxy fail it returns an html doc
-      raise MiqException::MiqInvalidCredentialsError, err.message, err.backtrace
```
Comment (enoodle) [inline app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/prometheus_client_mixin.rb]: Where is the credential error handled now? If it is in the client then line 55 will turn it intu `MiqUnreachableError`

File: app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/prometheus_client_mixin.rb (hunk 1 above)
Comment (enoodle) [reply app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/prometheus_client_mixin.rb]: Ok I see it in https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/195/files#diff-8ade5cbaa5f6435425cdbd87ed7d5a96R42

File: app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/prometheus_client_mixin.rb (hunk 2 above)
Comment (yaacov) [reply app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/prometheus_client_mixin.rb]: 👍 nice catch, now that we have a specialized client we can leverage it's error's to get better differentiations between specific types of errors.

Thanks @enoodle , will remove the code that hide the real error.

**Comment by moolitayer** [comment]
@yaacov FYI while reviewing this PR I noticed that we do not distinguish between different errors from this endpoint. ssl, unauthorized or non existing route all look the same. See for example ssl error against OpenShift and Prometheus:





This was not introduced here, we will want to improve that going forward

**Comment by cben** [comment]
Indeed.  The UI just show all failures the same:
https://github.com/ManageIQ/manageiq-ui-classic/blob/629f2d6/app/controllers/mixins/ems_common_angular.rb#L99-L100
There is some exception classification in AuthenticationMixin:
https://github.com/ManageIQ/manageiq/blob/ed4af9d204/app/models/mixins/authentication_mixin.rb#L348-L361
but it doens't separate out things like "we don't trust the server" from "server doesn't trust us".

What do you think better UX would look like?
And please open new issue if you want us not to forget this.

**Comment by bazulay** [comment]
@moolitayer  please open a github issue based on the 2 previous comments

**Comment by moolitayer** [comment]
https://github.com/ManageIQ/manageiq-providers-kubernetes/issues/200

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit 6fe43de2c8495d3502d019a29fe6d3b9eb256df6
Author: Mooli Tayer <mtayer@redhat.com>
Date:   Wed Dec 27 18:00:01 2017 +0200

    Merge pull request #195 from yaacov/use-prometheus-client
    
    Use prometheus client instead of faraday
    (cherry picked from commit 37b7884b16ea879764bac680f8bd2b60c3c9c3a3)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1530736
```

**Pull request merged**